modules/javafx.graphics/src/main/java/com/sun/javafx/css/BinarySerializer.java

```diff
@@ -0,0 +1,186 @@
+import java.util.Set;
+
+import javafx.css.PseudoClass;
+import javafx.css.Selector;
+import javafx.css.StyleConverter;
+import javafx.geometry.NodeOrientation;
+
+import static javafx.geometry.NodeOrientation.LEFT_TO_RIGHT;
+import static javafx.geometry.NodeOrientation.RIGHT_TO_LEFT;
+
+/**
+ * Class which can read and write selectors in a binary format.
+ */
+public class BinarySerializer {
+    private static final int TYPE_SIMPLE = 1;
+    private static final int TYPE_COMPOUND = 2;
+
+    public static Selector read(DataInputStream is, String[] strings) throws IOException {
+        int type = is.readByte();
+
+        if (type == TYPE_SIMPLE) {
+            return readSimpleSelector(is, strings);
+        }
+
+        // Backwards compatible, if it is not TYPE_SIMPLE any other value is considered TYPE_COMPOUND
+        return readCompoundSelector(is, strings);
+    }
+
+    public static void write(Selector selector, DataOutputStream os, StyleConverter.StringStore stringStore) throws IOException {
+        if (selector instanceof SimpleSelector s) {
+            writeSimpleSelector(s, os, stringStore);
+        }
+        else if(selector instanceof CompoundSelector s) {
+            writeCompoundSelector(s, os, stringStore);
+        }
+        else {
+            throw new IllegalStateException("support missing for selector type: " + (selector == null ? "null" : selector.getClass()));
+        }
+    }
+
+    private static SimpleSelector readSimpleSelector(DataInputStream is, String[] strings) throws IOException {
+        String name = strings[is.readShort()];
+        int nStyleClasses = is.readShort();
+        List<String> styleClasses = new ArrayList<>();
+
+        for (int n = 0; n < nStyleClasses; n++) {
+            styleClasses.add(strings[is.readShort()]);
+        }
+
+        String id = strings[is.readShort()];
+        int nPseudoclasses = is.readShort();
+        List<String> pseudoclasses = new ArrayList<>();
+
+        for (int n = 0; n < nPseudoclasses; n++) {
+            pseudoclasses.add(strings[is.readShort()]);
+        }
+
+        return new SimpleSelector(name, styleClasses, pseudoclasses, id);
+    }
+
+    private static CompoundSelector readCompoundSelector(DataInputStream is, String[] strings) throws IOException {
+        int nSelectors = is.readShort();
+        List<SimpleSelector> selectors = new ArrayList<>();
+
+        for (int n = 0; n < nSelectors; n++) {
+
+            /*
+             * An extra byte is part of the binary format containing TYPE_SIMPLE that
+             * isn't strictly needed. However, to remain backwards compatible with the
+             * first (and only) supported binary format, this byte must be skipped over
+             * before attempting to read what is known to be a simple selector.
+             */
+
+            is.readByte();
+
+            selectors.add(readSimpleSelector(is, strings));
+        }
+
+        int nRelationships = is.readShort();
```

Comment:
same here: should we check for a positive value?

as a general rule, we should be validating the input as it might come from untrusted sources, right?  L79 and other places?

Reply:
I considered doing more here, but as this is all just moved code, I'm hesitant to change it as part of this PR.  For example, if there is a faulty binary CSS file which has a negative value for the short, then the original code will just skip the loop.  If I add a check, it will change the behavior.

Reply:
Well, we could create a separate ticket.

Reply:
I already cleaned it up without altering the behavior.  It probably isn't worth investing more time in.  The whole binary CSS code is AFAICS basically only used to create binary versions of the FX standard css files, like `modena.css`, by calling the main method of `Css2Bin` during the gradle build.  This probably saves some time for JavaFX startup as the binary version will load faster than the text versions (or so I hope, I didn't check if that's true).  I have doubts it is used anywhere else, so although this has creeped into a few public API's, it's practically for internal use only.